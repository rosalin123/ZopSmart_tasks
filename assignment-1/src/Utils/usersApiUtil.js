const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchUsers = async () => {
  try {
    let url = `${baseUrl}/users`;
    let users = await fetch(url, { method: 'GET' });
    let usersJson = await users.json();
    return usersJson;
  } catch (err) {
    console.log('Error fetching users details', err.stack);
  }
};

export const getUser = async (id) => {

    let url = `${baseUrl}/users/${id}`;
    let user = await fetch(url, { method: 'GET' });
    let userJson = await user.json();
    return userJson;

};
