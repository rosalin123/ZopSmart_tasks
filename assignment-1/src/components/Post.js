import React, { Component } from 'react';
import { getPost } from '../actions/postActions';
import { fetchUsers } from '../actions/userActions';
import { fetchPostComments } from '../actions/postActions';

import { connect } from 'react-redux';

class Post extends Component {
  componentDidMount = async () => {
    this.props.fetchUsers();
    this.props.getPost();
    this.props.fetchPostComments();
  };

  render() {
    console.log('comments', this.props.comments);
    return (
      <div>
        <h3>{this.props.post.title}</h3>
        {this.props.users
          .filter((user) => user.id === this.props.post.userId)
          .map((user, index) => {
            return <h1 key={index}>{user.username}</h1>;
          })}

        <p>{this.props.post.body}</p>

        <h1>Comments</h1>
        {this.props.comments.map((comment, index) => {
          return (
            <div key={index}>
              <h3>{comment.name}</h3>
              <p>{comment.body}</p>
            </div>
          );
        })}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  post: state.Posts.item,
  users: state.Users.items,
  comments: state.Comments,
});

export default connect(mapStateToProps, {
  getPost,
  fetchUsers,
  fetchPostComments,
})(Post);
