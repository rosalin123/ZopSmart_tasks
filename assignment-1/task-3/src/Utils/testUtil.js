import React from 'react';

export const findbyTestAttr = (component, attr) => {
  const wrapper = component.find(`[data-test='${attr}']`);
  return wrapper;
};
