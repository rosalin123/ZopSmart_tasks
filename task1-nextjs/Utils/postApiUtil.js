import axios from 'axios';
import fetch from 'isomorphic-unfetch';
const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchPosts = async () => {
  let url = `${baseUrl}/posts`;
  let posts = await axios.get(url);
  let postsJson = await posts.json();
  return postsJson;
};

export const getPost = async (id) => {
  let url = `${baseUrl}/posts/${id}`;
  let post = fetch(url);
  let postJson = await post.json();
  return postJson;
};

export const getPostComments = async (id) => {
  let url = `${baseUrl}/posts/${id}/comments`;
  let comments = await axios.get(url);
  let commentsJson = await comments.json();
  return commentsJson;
};

export const getUserPosts = async (id) => {
  let url = `${baseUrl}/posts?userId=${id}`;
  let userPosts = await fetch(url);
  let userPostsJson = await userPosts.json();
  return userPostsJson;
};
