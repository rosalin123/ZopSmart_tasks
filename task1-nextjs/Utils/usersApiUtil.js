import fetch from 'isomorphic-unfetch';
const baseUrl = 'https://jsonplaceholder.typicode.com';

export const fetchUsers = async () => {
  try {
    let url = `${baseUrl}/users`;
    let users = await fetch(url);
    let usersJson = await users.json();
    return usersJson;
  } catch (err) {
    console.log('Error fetching users details', err.stack);
  }
};

export const getUser = async (id) => {
  let url = `${baseUrl}/users/${id}`;
  let user = await fetch(url);
  let userJson = await user.json();
  return userJson;
};
