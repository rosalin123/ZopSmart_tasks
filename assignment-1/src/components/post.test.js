import React from 'react';
import { shallow, render, mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import sinon from 'sinon';
import { BrowserRouter as Router } from 'react-router-dom';
import Post, { Post as PostComponent } from './Post';
import { mapDispatchToProps } from './Post';
import { findbyTestAttr, findByAltText } from '../Utils/testUtil';
import { middleware } from '../store';

const mockStore = configureMockStore(middleware);

const setUp = (initialState = {}) => {
  const store = mockStore(initialState);
  const component = render(
    <Router>
      <Provider store={store}>
        <Post />
      </Provider>
    </Router>
  );
  return component;
};

describe('Post component', () => {
  let component;

  it('should render without throwing any error', () => {
    const initialState = {
      Post: {
        loading: false,
        post: { title: 'abc', body: 'def' },
        error: '',
      },
      User: {
        loading: false,
        user: { id: 1, name: 'rose', email: 'r@gmail.com' },
        error: '',
      },
      Comments: {
        loading: false,
        comments: [{ name: 'abc', body: 'abc', email: 'abc' }],
        error: '',
      },
    };
    component = setUp(initialState);

    expect(component).toMatchSnapshot();
    expect(findbyTestAttr(component, 'PostComponent').length).toBe(1);
    expect(findbyTestAttr(component, 'loader').length).toBe(0);
  });

  it('should render Loader when loading is true', () => {
    const initialState = {
      Post: { loading: true, error: '' },
      User: { loading: false, error: '' },
      Comments: { loading: false, error: '' },
    };
    component = setUp(initialState);
    expect(findbyTestAttr(component, 'loader').length).toBe(1);
  });

  it('should render ErrorComponent when error occurs', () => {
    const initialState = {
      Post: { loading: false, post: [], error: 'Something wrong' },
      User: { loading: false, user: {}, error: '' },
      Comments: { loading: false, comments: [], error: '' },
    };
    component = setUp(initialState);

    let wrapper = findbyTestAttr(component, 'errorComponent');
    expect(wrapper.length).toBe(1);
    expect(findbyTestAttr(component, 'loader').length).toBe(0);
  });

  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).getPost(1);
    mapDispatchToProps(dispatch).getUser(1);
    mapDispatchToProps(dispatch).fetchPostComments(1);
    expect(typeof dispatch.mock.calls[0][0]).toBe('function');
    expect(typeof dispatch.mock.calls[1][0]).toBe('function');
    expect(typeof dispatch.mock.calls[2][0]).toBe('function');
  });

  it('componentDidMount calls the methods as expected', () => {
    const props = {
      match: { params: { id: 1 } },
      getPost: jest.fn(),
      getUser: jest.fn(),
      fetchPostComments: jest.fn(),
      postLoading: false,
      commentsLoading: false,
      postError: '',
      commentsError: '',
      userLoading: false,
      userError: '',
      post: [],
      user: {},
      comments: [],
    };

    component = shallow(<PostComponent {...props} />);
    const instance = component.instance();

    jest.spyOn(instance.props, 'getPost');
    jest.spyOn(instance.props, 'fetchPostComments');
    instance.componentDidMount();
    expect(instance.props.getPost).toHaveBeenCalledTimes(1);
    expect(instance.props.fetchPostComments).toHaveBeenCalledTimes(1);
  });
});
