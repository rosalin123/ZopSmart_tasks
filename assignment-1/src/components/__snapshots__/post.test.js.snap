// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component should render without throwing any error 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    data-test="PostComponent"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
        data-test="postDetails"
      >
        <div
          class="MuiCardHeader-root"
          data-test="postHeader"
        >
          <div
            class="MuiCardHeader-avatar"
          >
            <div
              aria-label="recipe"
              class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-7 MuiAvatar-colorDefault"
            />
          </div>
          <div
            class="MuiCardHeader-content"
          >
            <span
              class="MuiTypography-root MuiCardHeader-title MuiTypography-body2 MuiTypography-displayBlock"
            >
              <p
                class="makeStyles-cardHeaderStyles-3"
                style="margin:0;padding:0"
              />
            </span>
            <span
              class="MuiTypography-root MuiCardHeader-subheader MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
            >
              September 14, 2016
            </span>
          </div>
          <div
            class="MuiCardHeader-action"
          >
            <button
              aria-label="settings"
              class="MuiButtonBase-root MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
        <div
          class="MuiCardContent-root"
          data-test="postBody"
        >
           
          <a
            class="makeStyles-linkStyles-8"
            data-test="postLink"
            href="/posts/undefined"
          >
            <p
              class="MuiTypography-root makeStyles-cardContentTitleStyles-4 MuiTypography-body2 MuiTypography-colorTextPrimary"
              data-test="postTitle"
            >
              abc
            </p>
          </a>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            data-test="postContent"
          >
            def
          </p>
          <div
            class="MuiGrid-root MuiGrid-container"
            data-test="postComments"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <p
                class="MuiTypography-root makeStyles-cardContentTitleStyles-4 MuiTypography-body1"
              >
                 Comments 
              </p>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <div
                class="makeStyles-root-9"
                data-test="comment"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-align-items-xs-center"
                  data-test="avatar"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <div
                      class="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-12 MuiAvatar-colorDefault"
                    >
                      a
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item"
                  >
                    <p
                      class="MuiTypography-root makeStyles-usernameStyles-11 MuiTypography-body1"
                    >
                      abc
                    </p>
                  </div>
                </div>
                <h4
                  data-test="name"
                >
                  abc
                </h4>
                <p
                  class="MuiTypography-root makeStyles-typographyStyles-10 MuiTypography-body1 MuiTypography-colorTextSecondary"
                  data-test="body"
                >
                  abc
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiCardActions-root"
        >
          <button
            aria-label="add to favorites"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="share"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81 1.66 0 3-1.34 3-3s-1.34-3-3-3-3 1.34-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9c-1.66 0-3 1.34-3 3s1.34 3 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.16c-.05.21-.08.43-.08.65 0 1.61 1.31 2.92 2.92 2.92 1.61 0 2.92-1.31 2.92-2.92s-1.31-2.92-2.92-2.92z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-expanded="false"
            aria-label="show more"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-5"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
